pypesto/objective/finite_difference.py

```diff
@@ -1,13 +1,215 @@
+        function.
+        FDDelta.CONSTANT means that the step size is only initially selected.
+        FDDelta.SPACE means that the step size is updated if the current
+        evaluation point is sufficiently far away from the last training point.
+        FDDelta.STEPS means that the step size is updated `max_steps`
+        evaluations after the last update.
+    max_steps:
+        Number of steps after which to update in the `FDDelta.STEPS` update
+        condition.
+    """
+
+    # update conditions
+    CONSTANT = "constant"
+    SPACE = "space"
+    STEPS = "steps"
+    UPDATE_CONDITIONS = [CONSTANT, SPACE, STEPS]
+
+    def __init__(
+        self,
+        delta: Union[np.ndarray, float, None] = None,
+        test_deltas: np.ndarray = None,
+        update_condition: str = CONSTANT,
+        max_steps: int = 30,
+    ):
+        if not isinstance(delta, (np.ndarray, float)) and delta is not None:
+            raise ValueError(f"Unexpected type {type(delta)} for delta")
+        self.delta: Union[np.ndarray, float, None] = delta
+
+        if test_deltas is None:
+            test_deltas = np.array([10**(-i) for i in range(2, 9)])
+        self.test_deltas: np.ndarray = test_deltas
+
+        if update_condition not in FDDelta.UPDATE_CONDITIONS:
+            raise ValueError(
+                f"Update condition {update_condition} must be in "
+                f"{FDDelta.UPDATE_CONDITIONS}.",
+            )
+        self.update_condition: str = update_condition
+
+        self.max_steps: int = max_steps
+
+        # running variables
+        #  parameter where the step sizes where updated last
+        self.x0: Union[np.ndarray, None] = None
+        #  overall number of steps
+        self.steps: int = 0
+
+    def update(
+        self,
+        x: np.ndarray,
+        fval: Union[float, np.ndarray, None],
+        fun: Callable,
+        fd_method: str,
+    ) -> None:
+        """Update delta if update conditions are met.
+
+        Parameters
+        ----------
+        x:
+            Current parameter vector, shape (n_par,).
+        fval:
+            fun(x), to avoid re-evaluation. Scalar- or vector-valued.
+        fun:
+            Function whose 1st-order derivative to approximate.
+            Scalar- or vector-valued.
+        fd_method:
+            FD method employed by
+            :class:`pypesto.objective.finite_difference.FD`, see there.
+        """
+        # scalar to vector
+        if isinstance(self.delta, float):
+            self.delta: np.ndarray = self.delta * np.ones(shape=x.shape)
+
+        self.steps += 1
+
+        # return if no update needed
+        if self.delta is not None:
+            if (
+                self.update_condition == FDDelta.SPACE and
+                np.sum((x - self.x0)**2) <= 0.5 * np.sqrt(len(x))
+            ):
+                return
+            if (
+                self.update_condition == FDDelta.STEPS and
```

Comment:
I might be wrong, but as `self.steps==0` initially, update condition `STEPS` would update the deltas at the first update making the initial deltas obsolete?

Reply:
true, added "and self.steps > 1"